### PR TITLE
[v5] Inspect API

### PR DESCRIPTION
**Pull request by davidkpiano**

### CODE REVIEW

File: packages/core/test/inspect.test.ts [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+          type: event.type,
+          sessionId: event.sessionId,
+          snapshot:
+            typeof event.snapshot === 'object' && 'value' in event.snapshot
+              ? { value: event.snapshot.value }
+              : event.snapshot,
+          event: event.event.type
+        };
+      })
+    ).toMatchInlineSnapshot(`
```
Comment (davidkpiano) [inline packages/core/test/inspect.test.ts:160]: cc. @gavination Here are what the events look like, roughly

File: packages/core/src/stateUtils.ts [hunk 2]
```diff
@@ -1434,26 +1438,36 @@ export function resolveActionsAndContext<
       action: typeof action === 'string' ? { type: action } : (action as any)
     };
 
-    if (!('resolve' in resolved)) {
+    if (!('resolve' in resolvedAction)) {
       if (actorCtx?.self.status === ActorStatus.Running) {
-        resolved(args);
+        resolvedAction(args);
+        actorCtx?.system?._sendInspectionEvent({
```
Comment (rlaffers) [inline packages/core/src/stateUtils.ts:1444]: The ActorActionEvent is very useful for observing side effects performed by the actor. However, there is not enough information here. It is impossible to associate this action with the transition on which it was executed. Now is it possible to glean which actor performed this event. We might need the following:

```
{
  type: '@xstate.action`,
  sessionId: string, // <-- which actor executed this?
  event: AnyEventObject, // <-- as in ActorTransitionEvent.event
}
```

File: packages/core/src/system.ts [hunk 3]
```diff
@@ -34,6 +46,36 @@ export function createSystem<T extends ActorSystemInfo>(): ActorSystem<T> {
...
+    },
+    inspect: (observer) => {
+      observers.add(observer);
+    },
+    _sendInspectionEvent: (event) => {
+      const resolvedInspectionEvent: ResolvedInspectionEvent = {
+        id: Math.random().toString(),
+        createdAt: new Date().toString(),
+        ...event,
+        systemId: system.root.sessionId
```
Comment (rlaffers) [inline packages/core/src/system.ts:58]: I find this a little confusing. It is called systemId, but is populated with the root sessionId. 

I understand that this is meant to identify the system within which this is happening, but the `systemId` prop is already used in the spawn and invoke options. A child machine can have both sessionId and systemId, which may have been assigned to it when it was spawned/invoked.
I guess, I'm trying to say that Actor.systemId has a different meaning than InspectionEvent.systemId. 

- Actor.systemId = a key in the system Map,
- InspectionEvent.systemId = sessionId of the system itself.

**Review by rlaffers — COMMENTED** [review]
Great work, this will make implementing devtools much easier than in v4. :heart: 

Also, I have tested how it works - mostly very well.

One hitch is delayed actions - they are logged at the time their timer runs out, which makes sense. The inspector however still may wish to map them onto the transition which originally scheduled them. This is currently impossible, but adding an `event` prop onto the @xstate.action + a timestamp at which they were scheduled would allow for this mapping.

File: packages/core/src/system.ts (hunk 3 above)
Comment (davidkpiano) [reply packages/core/src/system.ts:58]: Yeah I struggled with this too, maybe we should rename this to `rootSessionId`

Comment (rlaffers) [reply packages/core/src/system.ts:58]: Yes, I think rootSessionId is a good name.

File: packages/core/src/stateUtils.ts (hunk 2 above)
Comment (rlaffers) [reply packages/core/src/stateUtils.ts:1444]: Additionally, an indication whether this was an entry, exit, or a transition action would be nice. It would allow for a better visualization in the Inspector.

File: packages/core/src/actions/send.ts [hunk 4]
```diff
@@ -123,10 +123,12 @@ function execute(
   const { to, event } = params;
 
   actorContext.defer(() => {
-    to.send(
+    actorContext?.system.sendTo(
```
Comment (davidkpiano) [inline packages/core/src/actions/send.ts:126]: By using `system.sendTo`, we make it possible to track the source/target of events.

File: packages/core/src/interpreter.ts [hunk 5]
```diff
@@ -439,6 +478,14 @@ export class Actor<
         `Only event objects may be sent to actors; use .send({ type: "${event}" }) instead`
       );
     }
+    if (!('__id' in event)) {
```
Comment (davidkpiano) [inline packages/core/src/interpreter.ts:481]: The `__id` is a hacky workaround to ensure that the same event isn't inspected twice

File: packages/core/test/inspect.test.ts [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+    const events: InspectionEvent[] = [];
+
+    const actor = interpret(machine, {
+      inspect: {
+        next(event) {
+          events.push(event);
+          // push events to websocket server
+          // server.clients.forEach((client) => {
+          //   client.send(JSON.stringify(event));
+          // });
```
Comment (davidkpiano) [inline packages/core/test/inspect.test.ts:70]:
```suggestion
```

File: packages/xstate-inspect/src/index.ts [hunk 7]
```diff
@@ -1,7 +1,114 @@
...
+
 export {
   inspect,
   createWindowReceiver,
   createWebSocketReceiver,
   createDevTools
 } from './browser.ts';
 export * from './types.ts';
+
+const WebSocket = require('ws');
```
Comment (davidkpiano) [inline packages/xstate-inspect/src/index.ts:18]: Ignore this file, will probably revert these changes

File: packages/core/src/actors/observable.ts [hunk 8]
```diff
@@ -45,9 +45,10 @@ export function fromObservable<T, TInput>(
   const errorEventType = '$$xstate.error';
   const completeEventType = '$$xstate.complete';
 
-  return {
+  // TODO: add event types
+  const logic: ObservableActorLogic<T, TInput> = {
```
Comment (Andarist) [inline packages/core/src/actors/observable.ts:49]: is there any need to move it to a variable? it's quite nice that we don't have to repeat the declared type here if we just rely on the contextual return type applied to the return statement

File: packages/core/src/actors/observable.ts [hunk 9]
```diff
@@ -45,9 +45,10 @@ export function fromObservable<T, TInput>(
   const errorEventType = '$$xstate.error';
   const completeEventType = '$$xstate.complete';
 
-  return {
+  // TODO: add event types
```
Comment (Andarist) [inline packages/core/src/actors/observable.ts:48]: which event types do u think about here?

File: packages/core/src/types.ts [hunk 10]
```diff
@@ -2154,11 +2156,31 @@ export interface ActorSystemInfo {
...
+   */
   _unregister: (actorRef: AnyActorRef) => void;
+  /**
+   * @private
+   */
   _set: <K extends keyof T['actors']>(key: K, actorRef: T['actors'][K]) => void;
   get: <K extends keyof T['actors']>(key: K) => T['actors'][K] | undefined;
+  inspect: (observer: Observer<InspectionEvent>) => void;
+  _sendInspectionEvent: (event: InspectionEvent) => void;
+  sendTo: (
```
Comment (Andarist) [inline packages/core/src/types.ts:2178]: I like this method but I wonder about:
- could we use a different order of arguments, having `event` at the end would look visually better when we call this function
- could we make `target` and `source` required?

File: packages/core/src/types.ts [hunk 11]
```diff
@@ -2154,11 +2156,31 @@ export interface ActorSystemInfo {
 }
 
 export interface ActorSystem<T extends ActorSystemInfo> {
+  /**
+   * @private
+   */
```
Comment (Andarist) [inline packages/core/src/types.ts:2161]: A neat thing that we could maybe utilize is `@internal` and TS `stripInternal: true` option

File: packages/core/src/actions/send.ts [hunk 12]
```diff
@@ -123,10 +123,12 @@ function execute(
   const { to, event } = params;
 
   actorContext.defer(() => {
```
Comment (Andarist) [inline packages/core/src/actions/send.ts:125]: I wonder if we could move this concern to the `system.sendTo`. It would be nice if we didn't have to worry about this here at all and if all sends would be processed in the very same way. That's outside of the scope of this PR though

File: packages/core/src/interpreter.ts [hunk 13]
```diff
@@ -83,6 +87,8 @@ export class Actor<
   TEvent extends EventObject = EventFromLogic<TLogic>
 > implements ActorRef<TEvent, SnapshotFrom<TLogic>>
 {
+  public static defaults = defaultOptions as Partial<ActorOptions<any>> &
```
Comment (Andarist) [inline packages/core/src/interpreter.ts:90]: Why do we expose this as a static property here? It feels like somebody could mutate this to change the default but I would consider that to be a bad thing overall. I wouldn't like a random~ call site to meaningfully change the behavior of all actors in the application. While it might make sense as something convenient in a self-contained system - it quickly breaks apart situations in which actors from different subsystems are put in a single application. A change "intended" for one of them should really not affect **all of them**.

I understand that this might be used to enable devtools globally but I would much prefer a dedicated way of doing this. Inspection is an extra thing that doesn't affect runtime behaviors of all contained subsystems so I think that setting this one thing globally should be fine.

Comment (davidkpiano) [reply packages/core/src/interpreter.ts:90]: What would be the dedicated way of enabling devtools globally that you're thinking of?

File: packages/core/src/interpreter.ts [hunk 14]
```diff
@@ -135,14 +141,18 @@ export class Actor<
...
+    } as ActorOptions<TLogic> & typeof Actor.defaults;
+
+    const { clock, logger, parent, id, systemId, inspect } = resolvedOptions;
 
-    const { clock, logger, parent, id, systemId } = resolvedOptions;
-    const self = this;
+    this.system = parent?.system ?? createSystem(this);
 
-    this.system = parent?.system ?? createSystem();
+    if (inspect) {
```
Comment (Andarist) [inline packages/core/src/interpreter.ts:152]: I think there is no way atm to pass `inspect` when invoking or spawning children but perhaps better be safe than sorry here?
```suggestion
    if (inspect && !parent) {
```

File: packages/core/src/interpreter.ts (hunk 13 above)
Comment (Andarist) [reply packages/core/src/interpreter.ts:90]:
```ts
import { enableInspection } from 'xstate/devtools'
enableInspection(myInspector)
```

In the future we could even allow this:
```ts
import 'xstate/devtools/enable-browser-devtools'
```

Internally, we would just set a module-level variable and reach for it when instantiating a system.

File: packages/core/src/types.ts (hunk 10 above)
Comment (davidkpiano) [reply packages/core/src/types.ts:2223]: Sure, we can reorder it. It should be an internal-only method too, not meant for public usage.

File: packages/core/src/actors/observable.ts (hunk 8 above)
Comment (davidkpiano) [reply packages/core/src/actors/observable.ts:49]: No need, can revert

File: packages/core/src/types.ts [hunk 15]
```diff
@@ -2219,3 +2241,55 @@ export type PersistedMachineState<TState extends AnyState> = Pick<
     };
   };
 };
+
+export interface InspectedActorObject {
```
Comment (Andarist) [inline packages/core/src/types.ts:2245]: it would be nice if all of those inspector-related types would be moved closer to where they are actually used - which currently is the `system.ts`. I find it much easier to navigate the codebase when things stay colocated. Our `./types.ts` here is pretty huge already and I often get lost in it ;p

File: packages/core/src/types.ts (hunk 10 above)
Comment (Andarist) [reply packages/core/src/types.ts:2223]: > It should be an internal-only method too, not meant for public usage.

This has crossed my mind but it actually should be available for all custom actors so we can't just `_` it

File: packages/core/src/types.ts [hunk 16]
```diff
@@ -2219,3 +2241,55 @@ export type PersistedMachineState<TState extends AnyState> = Pick<
...
+  sourceId: string | undefined; // Session ID
+  targetId: string; // Session ID, required
+}
+
+export interface ActorRegistrationEvent {
+  type: '@xstate.actor';
+  actorRef: AnyActorRef;
+  sessionId: string;
+  parentId?: string;
+  definition?: string; // JSON-stringified definition or URL
```
Comment (Andarist) [inline packages/core/src/types.ts:2285]:
```suggestion
```

File: packages/core/src/interpreter.ts [hunk 17]
```diff
@@ -179,6 +189,13 @@ export class Actor<
     // Ensure that the send method is bound to this Actor instance
     // if destructured
     this.send = this.send.bind(this);
+    this.system._sendInspectionEvent({
+      type: '@xstate.actor',
+      actorRef: this,
+      sessionId: this.sessionId,
+      // definition: JSON.stringify(this.logic.config),
```
Comment (Andarist) [inline packages/core/src/interpreter.ts:196]:
```suggestion
```

File: packages/core/src/interpreter.ts [hunk 18]
```diff
@@ -193,7 +210,7 @@ export class Actor<
   // array of functions to defer
   private _deferred: Array<() => void> = [];
 
-  private update(state: InternalStateFrom<TLogic>): void {
+  private update(state: InternalStateFrom<TLogic>, event: TEvent): void {
```
Comment (Andarist) [inline packages/core/src/interpreter.ts:213]: this would likely save you some casts:
```suggestion
  private update(state: InternalStateFrom<TLogic>, event: EventObject): void {
```

`TEvent` annotation isn't true here anyway since internal events don't belong to `TEvent`

File: packages/core/src/types.ts (hunk 15 above)
Comment (davidkpiano) [reply packages/core/src/types.ts:2245]: Will move 🚚

File: packages/core/src/interpreter.ts [hunk 19]
```diff
@@ -282,13 +311,22 @@ export class Actor<
     }
     this.status = ActorStatus.Running;
 
+    const initEvent = createInitEvent(this.options.input);
+
+    this.system._sendInspectionEvent({
+      type: '@xstate.event',
+      event: initEvent,
+      sourceId: this._parent?.sessionId,
```
Comment (Andarist) [inline packages/core/src/interpreter.ts:319]: I wonder if this is truly OK, is the init event received from the parent or is it raised internally? I can see arguments going both ways :p

File: packages/core/src/system.ts [hunk 20]
```diff
@@ -1,13 +1,25 @@
...
+): ActorSystem<T> {
   const children = new Map<string, AnyActorRef>();
   const keyedActors = new Map<keyof T['actors'], AnyActorRef | undefined>();
   const reverseKeyedActors = new WeakMap<AnyActorRef, keyof T['actors']>();
+  const observers = new Set<Observer<ResolvedInspectionEvent>>();
 
   const system: ActorSystem<T> = {
-    _bookId: () => `x:${sessionIdCounter++}`,
+    root: rootActor,
+    _bookId: (name: string = 'x') => `${name}:${systemCounter++}`,
```
Comment (Andarist) [inline packages/core/src/system.ts:22]: This argument is never used in the codebase
```suggestion
    _bookId: () => `x:${systemCounter++}`,
```

File: packages/core/src/types.ts [hunk 21]
```diff
@@ -2219,3 +2241,53 @@ export type PersistedMachineState<TState extends AnyState> = Pick<
...
+}
+
+export type InspectionEvent =
+  | ActorTransitionEvent
+  | ActorCommunicationEvent
+  | ActorRegistrationEvent;
+
+export type ResolvedInspectionEvent = InspectionEvent & BaseInspectionEvent;
+
+export type InspectorActorRef = ActorRef<InspectionEvent>;
```
Comment (Andarist) [inline packages/core/src/types.ts:2293]:
```suggestion
export type InspectorActorRef = ActorRef<InspectionEvent>;
```

Comment (Andarist) [inline packages/core/src/types.ts:2293]: such a thing doesn't exist anywhere so it's redundant, at least right now
```suggestion
```

File: packages/core/src/types.ts [hunk 22]
```diff
@@ -2219,3 +2241,53 @@ export type PersistedMachineState<TState extends AnyState> = Pick<
     };
   };
 };
+
+export interface InspectedActorObject {
```
Comment (Andarist) [inline packages/core/src/types.ts:2245]: this whole type is not used anywhere, can it be removed?

File: packages/core/src/types.ts [hunk 23]
```diff
@@ -2219,3 +2241,53 @@ export type PersistedMachineState<TState extends AnyState> = Pick<
...
+  systemId?: string; // Session ID
+  events: ActorCommunicationEvent[]; // events where targetId === sessionId
+  definition?: string; // JSON-stringified machine definition or URL
+  createdAt: number; // Timestamp
+  updatedAt: number; // Timestamp, derived from latest update createdAt
+  status: 0 | 1 | 2; // 0 = not started, 1 = started, 2 = stopped, derived from latest update status
+}
+
+export interface BaseInspectionEvent {
+  actorSystemId: string; // the session ID of the root
```
Comment (Andarist) [inline packages/core/src/types.ts:2258]: This one would sound a little bit more accurate. `actorSystemId` sounds a little bit like if we'd have special system ids but we don't - the system is attached to the root actor but it has no id of its own.
```suggestion
  rootId: string; // the session ID of the root
```

File: packages/core/src/types.ts [hunk 24]
```diff
@@ -2219,3 +2241,53 @@ export type PersistedMachineState<TState extends AnyState> = Pick<
...
+  type: '@xstate.actor';
+  actorRef: AnyActorRef;
+  sessionId: string;
+  parentId?: string;
+}
+
+export type InspectionEvent =
+  | ActorTransitionEvent
+  | ActorCommunicationEvent
+  | ActorRegistrationEvent;
```
Comment (Andarist) [inline packages/core/src/types.ts:2289]: All of those have somewhat confusing names. I think it would be better if the name of the type would match the `.type` closer - but currently, we have `ActorTransitionEvent` with `.type: '@xstate.snapshot'` etc.

File: packages/core/src/interpreter.ts [hunk 25]
```diff
@@ -282,13 +311,22 @@ export class Actor<
     }
     this.status = ActorStatus.Running;
 
+    const initEvent = createInitEvent(this.options.input);
+
+    this.system._sendInspectionEvent({
```
Comment (Andarist) [inline packages/core/src/interpreter.ts:316]: I wish that this one would go through `system.sendTo` if that's treated as an explicit event here. It's not a blocker though since ultimately it doesn't matter that much and we can try to juggle things more in the future.

File: packages/core/src/interpreter.ts (hunk 19 above)
Comment (davidkpiano) [reply packages/core/src/interpreter.ts:319]: Conceptually, parent creates a child with input that is inherited from the parent

File: packages/core/src/types.ts (hunk 22 above)
Comment (davidkpiano) [reply packages/core/src/types.ts:2245]: Yes

File: packages/core/src/interpreter.ts (hunk 13 above)
Comment (davidkpiano) [reply packages/core/src/interpreter.ts]: Un-exposed

File: packages/core/src/types.ts (hunk 10 above)
Comment (davidkpiano) [reply packages/core/src/types.ts]: Renamed to `relay(event, source, target)` as it is _relaying_ an event

File: packages/core/src/actions/send.ts (hunk 12 above)
Comment (davidkpiano) [reply packages/core/src/actions/send.ts:131]: Probably not easily at the moment; something we can revisit

File: packages/core/src/types.ts [hunk 26]
```diff
@@ -2176,11 +2181,30 @@ export interface ActorSystemInfo {
...
+   * @internal
+   */
   _unregister: (actorRef: AnyActorRef) => void;
+  /**
+   * @internal
+   */
   _set: <K extends keyof T['actors']>(key: K, actorRef: T['actors'][K]) => void;
   get: <K extends keyof T['actors']>(key: K) => T['actors'][K] | undefined;
+  inspect: (observer: Observer<InspectionEvent>) => void;
+  _sendInspectionEvent: (event: InspectionEvent) => void;
```
Comment (Andarist) [inline packages/core/src/types.ts:2202]:
```suggestion
  /**
   * @internal
   */
  _sendInspectionEvent: (event: InspectionEvent) => void;
```

**Review by Andarist — APPROVED** [review]
